### ISSUE TITLE
[BUG]: Attempting to unscale FP16 gradients

### ISSUE THREAD

**Issue by caolinzheng** [hpcaitech/ColossalAI#3736]
### 🐛 Describe the bug

I trained llama's model with colossal's engine, the error "Attempting to unscale FP16 gradients" occured while executing Engine.step (). 
The training environment is as follows：GPU:V100*4，tp_degree=4，data_degree=1
config.py：
from colossalai.amp import AMP_TYPE
fp16=dict(
    mode=AMP_TYPE.TORCH
)
parallel=dict(
    data=1,
    pipeline=1,
    tensor=dict(size=4, mode='1d'),
)
gradient_accumulation=64
clip_grad_norm = 1.0
BATCH_SIZE = 2
NUM_EPOCHS = 3

training code:
 with ctx:
        model = LlamaForCausalLM(config)
        for n, p in model.named_parameters():
            x = state_dict[n]
            x = x.chunk(tp_degree, dim=-1)
            part = rank_local % tp_degree
            x = x[part]
            p.data.copy_(x)
    if tp_degree > 1:
        tensor_parallelize(model, tp_pg)
    optimizer = HybridAdam(model.parameters(), lr=1e-5)
    criterion = LLamaLoss()
    cuda_device = torch.device("cuda")
    cpu_device = torch.device("cpu")
    model = model.to(cuda_device)
    data_collator = DataCollatorForSupervisedDataset(tokenizer=tokenizer)

    train_dataset = SupervisedDataset(tokenizer=tokenizer, data_path="alpaca_data.json.all")
    test_dataset = SupervisedDataset(tokenizer=tokenizer, data_path="data.100")

    train_sampler = torch.utils.data.distributed.DistributedSampler(dataset=train_dataset, num_replicas=tp_pg._dp_degree, rank=int(rank_local/tp_pg._tp_degree))
    test_sampler = torch.utils.data.distributed.DistributedSampler(dataset=test_dataset, num_replicas=tp_pg._dp_degree, rank=int(rank_local/tp_pg._tp_degree))

    train_dataloader = DataLoader(train_dataset, batch_size=gpc.config.BATCH_SIZE, sampler=train_sampler, collate_fn=data_collator)
    test_dataloader = DataLoader(train_dataset, batch_size=gpc.config.BATCH_SIZE, sampler=test_sampler, collate_fn=data_collator)
    engine, train_dataloader, test_dataloader, _ = colossalai.initialize(model,
                                                                                                                   optimizer,
                                                                                                                   criterion,
                                                                                                                   train_dataloader,
                                                                                                                   test_dataloader)

    print(f"rank={rank_local} dp_rank_list={tp_pg._dp_rank_list}")
    print(f"rank={rank_local} tp_rank_list={tp_pg._tp_rank_list}")
    engine.train()

    for e in range(gpc.config.NUM_EPOCHS):
        train_sampler.set_epoch(e)
        for i, data in enumerate(train_dataloader):
            engine.zero_grad()
            torch.cuda.synchronize()
            start = time.time()
            labels = data["labels"].to(get_current_device())
            input_ids = data["input_ids"].to(get_current_device())
            attention_mask = data["attention_mask"].to(get_current_device())
            #output = model(input_ids=input_ids, labels=labels, attention_mask=attention_mask)
            output = engine(input_ids=input_ids, attention_mask=attention_mask)
            logits = output["logits"]
            loss = engine.criterion(logits, labels, acc_step=gpc.config.gradient_accumulation, tp_degree=tp_degree, vocab_size=config.vocab_size)
            #optimizer.backward(loss)
            engine.backward(loss)
            engine.step()
            engine.zero_grad()
            torch.cuda.synchronize()
            end = time.time()
            avg_item_time = (end - start)/(gpc.config.BATCH_SIZE*tp_pg._dp_degree)
            logger.info(
                f"Loss:{loss.item():.3f}, Step time: {avg_item_time:.3f}s/t input={input_ids.shape} idx={i+1}",
                ranks=[0]
            )

if __name__ == "__main__":
    train()
